### PR TITLE
Replace super-ensemble with probability blending

### PR DESCRIPTION
**Pull request by btrotta-bom**
# Summary

This PR replaces the super-ensemble methodology (as used in ECPoint) with a more mathematically justified approach for blending the probability distributions of the different ensemble members.

As an additional benefit, this reduces the runtime slightly and reduces memory usage by around 50%.

Testing:
 - [X] Ran tests and they passed OK
 - [X] Added new tests for the new feature(s)
 
Acceptance test data PR:
https://github.com/metoppv/improver_test_data/pull/7

# Detailed explanation of change

Currently, the rainforests model does the following:
1.	For each realization of the NWP ensemble, calculate the probabilities that the error exceeds certain pre-defined thresholds.
2.	For each ensemble member, extract from this distribution of error probabilities 20 evenly-spaced quantiles (0.05, 0.1, …).
3.	Combine the quantiles from each ensemble member into a super-ensemble. (Thus for a 50-member NWP ensemble, the super-ensemble has 1000 points.)
4.	Treat the members of the super-ensembles as random samples from the true distribution, and extract 100 percentiles from the super-ensemble as the output.

The final step is not mathematically justified. For example, consider a situation where all the NWP ensemble members are very close together, and the estimated error distributions for each member are similar, so that all the 0.05 quantile values are less than all the 0.1 quantile values, etc. In this case, the 0.95 quantile (call it t) of the super-ensemble will be the minimum of all the 0.95 quantiles of the individual error distributions. But this is a biased estimate, because, for all but one of the error distributions, the probability of exceeding t is strictly greater than 0.95. This PR implements the correct approach to combining probabilities, as described here https://github.com/metoppv/improver/files/1128018/Combining_Probabilities.pdf. We interpolate each ensemble member's distribution to the same set of output thresholds, then average over ensemble members.

### CODE REVIEW

File: improver/calibration/rainforest_calibration.py [hunk 1]
```diff
@@ -703,8 +601,8 @@ def process(
 
         2. Interpolate the CDF to extract a series of percentiles for the error distribution.
         The error percentiles are then applied to each associated ensemble realization to
-        produce a series of realisable values; collectively these series form a calibrated
-        super-ensemble, which is then sub-sampled to provide the calibrated forecast.
+        produce an error distribution. Distributions for different ensemble members are
+        then interpolated to the same set output thresholds and averaged.
```
Comment (benowen-bom) [inline improver/calibration/rainforest_calibration.py:605]: Would it be worth mentioning that they are averaged in probability space?

**Review by benowen-bom — COMMENTED** [review]
Thanks @btrotta-bom. The changes to the methodology look good and overall I think this solution is cleaner (both mathematically and in limiting variation in the final output through the removal of both the `error_percentiles_count` and `output_realizations_count`).

I've raised a few comments around the threshold arguments and have some suggested simplifications, but otherwise I think the code changes look good.

File: improver/cli/apply_rainforests_calibration.py [hunk 2]
```diff
@@ -68,15 +69,15 @@ def process(
...
-        output_realizations_count (int):
-            The number of realizations to output for the calibrated ensemble.
-            These realizations are sampled by taking equispaced percentiles
-            from the super-ensemble. If None is supplied, then all realizations
-            from the super-ensemble will be returned.
+        output_thresholds (list):
+            List of thresholds at which to evaluate output probabilties.
+        output_threshold_config (dict):
+            Threshold configuration dictionary where the keys are strings representing
+            thresholds and the values are all None.
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/cli/apply_rainforests_calibration.py:76]: Updated to clarify this issue.

File: improver/cli/apply_rainforests_calibration.py [hunk 3]
```diff
@@ -68,15 +69,15 @@ def process(
...
+        threshold_units (str):
+            Units in which output_thresholds or output_threshold_config are specified.
+            Will be converted to the units of  forecast_cube. If None, assumed to
+            be the same as forecast_cube.
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/cli/apply_rainforests_calibration.py:80]: Done.

File: improver/cli/apply_rainforests_calibration.py [hunk 4]
```diff
@@ -88,9 +89,23 @@ def process(
...
+            "--output-threshold-config and --output-thresholds are mutually exclusive "
+            "- please set one or the other, not both"
+        )
+    if (not output_threshold_config) and (not output_thresholds):
+        "One of --output-threshold-config and --output-thresholds must be specified"
+
+    if output_threshold_config:
+        thresholds = []
+        for key in output_threshold_config.keys():
+            thresholds.append(float(key))
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/cli/apply_rainforests_calibration.py:103]: done

File: improver/calibration/rainforest_calibration.py [hunk 5]
```diff
@@ -434,6 +429,7 @@ def _evaluate_probabilities(
 
         for threshold_index, model in enumerate(self.tree_models):
             threshold = self.error_thresholds[threshold_index]
+            # prediction = model.predict(input_dataset)
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py:432]: This was left over from testing, removed now.

File: improver/calibration/rainforest_calibration.py [hunk 6]
```diff
@@ -491,200 +488,101 @@ def _calculate_error_probabilities(
...
-        forecast_subensembles_data = np.maximum(0.0, forecast_subensembles_data)
-        # Return cube containing forecast subensembles
-        return create_new_diagnostic_cube(
-            name=forecast_cube.name(),
-            units=forecast_cube.units,
-            template_cube=error_percentiles_cube,
-            mandatory_attributes=generate_mandatory_attributes([forecast_cube]),
-            optional_attributes=forecast_cube.attributes,
-            data=forecast_subensembles_data,
+        probabilities = error_CDF.data
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py:527]: fixed.

File: improver/calibration/rainforest_calibration.py (hunk 1 above)
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py:605]: done

File: improver/calibration/rainforest_calibration.py [hunk 7]
```diff
@@ -491,200 +488,101 @@ def _calculate_error_probabilities(
...
+        num_forecast_dims = len(forecast.coords(dim_coords=True))
+        for i in range(num_forecast_dims):
+            error_values = np.expand_dims(error_values, axis=-1)
+        thresholds = error_values + np.expand_dims(forecast.data, axis=0)
+        probabilities = error_CDF.data
+
+        # Add bounds
+        epsilon = 0.0001
+        thresholds = np.concatenate(
+            [thresholds[[0]] - epsilon, thresholds, thresholds[[-1]] + epsilon], axis=0,
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: I've changed this to do something similar to the expected value cli, using the floating-point epsilon. I'm now using `get_bounds_of_distribution` (also changed in `_evaluate_probabilities`). I don't think `insert_lower_and_upper_endpoint_to_1d_array` will work here since `thresholds` is not a 1-d array.

File: improver/calibration/rainforest_calibration.py [hunk 8]
```diff
@@ -491,200 +488,101 @@ def _calculate_error_probabilities(
...
-            input_cube:
-                Cube containing the forecast_subensembles.
-
-        Returns:
-            Cube containing single realization dimension in place of the realization
-            and percentile dimensions in forecast_subensemble.
+        # Set probability to 1 for negative thresholds
+        negative_threshold = thresholds <= 0
+        thresholds = np.where(negative_threshold, 0, thresholds)
+        probabilities = np.where(negative_threshold, 1, probabilities)
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: Fixed.

File: improver/cli/apply_rainforests_calibration.py [hunk 9]
```diff
@@ -39,8 +39,9 @@ def process(
     forecast: cli.inputcube,
     *features: cli.inputcube,
     model_config: cli.inputjson,
-    error_percentiles_count: int = 19,
-    output_realizations_count: int = 100,
+    output_thresholds: cli.comma_separated_list_of_float = None,
+    output_threshold_config: cli.inputjson = None,
+    threshold_units: str = None,
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/cli/apply_rainforests_calibration.py:46]: I've updated the docstring to say that the cli assumes models have been trained using `>=` and output uses the same comparator.

File: improver/calibration/rainforest_calibration.py [hunk 10]
```diff
@@ -491,200 +488,101 @@ def _calculate_error_probabilities(
...
+            (coord.copy(), forecast.coord_dims(coord)[0] + 1)
+            for coord in forecast.coords(dim_coords=True)
+        ]
+        probability_cube = iris.cube.Cube(
+            output_array.astype(np.float32),
+            long_name=f"probability_of_{forecast_variable}_above_threshold",
+            units=1,
+            attributes=forecast.attributes,
             dim_coords_and_dims=dim_coords_and_dims,
             aux_coords_and_dims=aux_coords_and_dims,
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: Done

File: improver/calibration/rainforest_calibration.py [hunk 11]
```diff
@@ -491,200 +488,101 @@ def _calculate_error_probabilities(
...
+
+        # Add error to forecast
+        error_values = error_CDF.coord(
+            "forecast_error_of_lwe_thickness_of_precipitation_amount"
+        ).points
+        num_forecast_dims = len(forecast.coords(dim_coords=True))
+        for i in range(num_forecast_dims):
+            error_values = np.expand_dims(error_values, axis=-1)
+        thresholds = error_values + np.expand_dims(forecast.data, axis=0)
+        probabilities = error_CDF.data
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: I've added some more explanation, and simplified the code as suggested.

File: improver/calibration/rainforest_calibration.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py]: No worries. Thanks for doing. As to the `eps` approach, I can understand the reasoning however note that if the `error_thresholds` don't provide suitable bounding values, then it may be possible that the probabilities are interpolated over a VERY small gap. That being said, I appreciate that specifying a bound (such as upper ECC bound) may result in interpolations over a very large gap. I'm not sure there is a good solution here other than to say that the user should ensure they have suitable limits for the upper and lower `error_thresholds`.

**Review by benowen-bom — COMMENTED** [review]
A few minor comments, but other than that it's looking good. One thing was that the acceptance tests are failing at the moment, but I think this should be fixed with a rebase to main on https://github.com/metoppv/improver_test_data/pull/7. I won't approve until the acceptance tests are all passing, so let me know once this has been fixed and I'll have another look.

File: improver/cli/apply_rainforests_calibration.py [hunk 12]
```diff
@@ -68,15 +75,17 @@ def process(
...
-            These realizations are sampled by taking equispaced percentiles
-            from the super-ensemble. If None is supplied, then all realizations
-            from the super-ensemble will be returned.
+        output_thresholds (list):
+            List of thresholds at which to evaluate output probabilties.
+        output_threshold_config (dict):
+            Threshold configuration dictionary where the keys are strings representing
+            thresholds. The dictionary values are not used and can be set to None (or
+            any other value). The interface to this cli is similar to that of the
+            threshold cli, but without the fuzzy bounds option.
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/cli/apply_rainforests_calibration.py]: done

File: improver/calibration/rainforest_calibration.py [hunk 13]
```diff
@@ -425,12 +421,10 @@ def _evaluate_probabilities(
...
-        lower_bound = bounds_data[0][0]
-        lower_bound_in_fcst_units = bounds_unit.convert(
-            lower_bound, forecast_variable_unit
+        bounds_data = get_bounds_of_distribution(
+            forecast_variable, forecast_variable_unit
         )
+        lower_bound_in_fcst_units = bounds_data[0]
 
         for threshold_index, model in enumerate(self.tree_models):
             threshold = self.error_thresholds[threshold_index]
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: done

File: improver/calibration/rainforest_calibration.py [hunk 14]
```diff
@@ -491,200 +486,107 @@ def _calculate_error_probabilities(
...
-        be the first and second dimensions respectively.
+        Add the error CDF to the NWP forecast and extract probabilities at output thresholds
+        for all realizations.
 
         Args:
-            input_cube:
-                Cube containing the forecast_subensembles.
+            error_CDF:
+                Cube containing the CDF of probabilities for each enemble member at error
+                threhsolds.
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: done

File: improver/calibration/rainforest_calibration.py [hunk 15]
```diff
@@ -491,200 +486,107 @@ def _calculate_error_probabilities(
...
-                The number of ensemble realizations that will be extracted from the
-                super-ensemble. If realizations_count is None, all realizations will
-                be returned.
-
-        Returns:
-            Cube containing single realization dimension.
-        """
-        superensemble_cube = self._stack_subensembles(forecast_subensembles)
+        # Set probability to 1 for thresholds outside bounds
+        impossible_threshold = thresholds <= lower_bound_in_fcst_units
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: Fair point, I've changed the variable name to make it clearer.

**Review by benowen-bom — DISMISSED** [review]
Thanks for making those final changes. I'm happy to sign off on this; I think the changes here are a good change to how the processing is done.

File: improver/ensemble_copula_coupling/utilities.py [hunk 16]
```diff
@@ -375,3 +376,60 @@ def interpolate_multiple_rows_same_y(*args):
...
+        fp: array with at least 2 dimensions. First dimension is the interpolation dimension.
+
+    Returns:
+        array with dimensions (len(x), ) + xp.shape[1:]
+    """
+
+    if xp.shape != fp.shape:
+        raise ValueError("xp and fp must have the same shape")
+    if len(xp.shape) < 2:
+        raise ValueError("xp and fp must have at least 2 dimensions")
```
Comment (gavinevans) [inline improver/ensemble_copula_coupling/utilities.py:398]: Codecov highlights that these lines aren't covered by any tests. It would be good to add a test to cover these lines and the duplicate lines in `fast_interpolate_pointwise`.

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks for these updates @btrotta-bom 👍 

I've added a few minor comments.

File: improver/cli/apply_rainforests_calibration.py [hunk 17]
```diff
@@ -68,15 +75,17 @@ def process(
...
-            The number of error percentiles to apply to each ensemble realization.
-            The resulting super-ensemble will be of size = forecast.realization.size *
-            error_percentiles_count.
-        output_realizations_count (int):
-            The number of realizations to output for the calibrated ensemble.
-            These realizations are sampled by taking equispaced percentiles
-            from the super-ensemble. If None is supplied, then all realizations
-            from the super-ensemble will be returned.
+        output_thresholds (list):
+            List of thresholds at which to evaluate output probabilties.
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/cli/apply_rainforests_calibration.py]: fixed

File: improver/cli/apply_rainforests_calibration.py [hunk 18]
```diff
@@ -88,9 +97,24 @@ def process(
 
     from improver.calibration.rainforest_calibration import ApplyRainForestsCalibration
 
+    if output_threshold_config and output_thresholds:
+        raise ValueError(
+            "--output-threshold-config and --output-thresholds are mutually exclusive "
+            "- please set one or the other, not both"
+        )
+    if (not output_threshold_config) and (not output_thresholds):
+        "One of --output-threshold-config and --output-thresholds must be specified"
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/cli/apply_rainforests_calibration.py]: Fixed, and added test.

File: improver/calibration/rainforest_calibration.py [hunk 19]
```diff
@@ -703,8 +607,8 @@ def process(
 
         2. Interpolate the CDF to extract a series of percentiles for the error distribution.
         The error percentiles are then applied to each associated ensemble realization to
-        produce a series of realisable values; collectively these series form a calibrated
-        super-ensemble, which is then sub-sampled to provide the calibrated forecast.
+        produce an error distribution. Distributions for different ensemble members are
+        then interpolated to the same set output thresholds and averaged in probability space.
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: fixed

File: improver/calibration/rainforest_calibration.py [hunk 20]
```diff
@@ -491,200 +488,107 @@ def _calculate_error_probabilities(
...
-            points=np.arange(realization_size * percentile_size, dtype=np.int32),
-            standard_name="realization",
-            units="1",
+
+        # Add error to forecast. The thresholds for each grid point are obtained
+        # by adding the forecast at that point to error_values; thus
+        # we get a different set of thresholds for each grid point.
+        error_values = error_CDF.coord(
+            "forecast_error_of_lwe_thickness_of_precipitation_amount"
+        ).points
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver/calibration/rainforest_calibration.py]: Good suggestion, thanks!

File: improver/ensemble_copula_coupling/utilities.py (hunk 16 above)
Comment (btrotta-bom) [reply improver/ensemble_copula_coupling/utilities.py:398]: Done

**Comment by btrotta-bom** [comment]
@gavinevans Thanks for your prompt review, I think I have addressed all your comments.

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks for the updates, @btrotta-bom 👍 

I've noticed a couple of accidental mistypes.

File: improver_tests/ensemble_copula_coupling/test_utilities.py [hunk 21]
```diff
@@ -661,6 +661,35 @@ def test_fast_interpolate_pointwise_called(self, interp_imp):
...
+
+    @skipIf(not (numba_installed), "numba not installed")
+    def test_shape_errors_fast(self):
+        """Test that an error is raised if inputs have the wrong shape."""
+        x = np.arange(1, 5)
+        xp = np.ones((2, 3))
+        fp = np.ones((3, 2))
+        msg = r"xp and fp must have the same shape"
+        with self.assertRaisesRegex(ValueError, msg):
+            slow_interpolate_pointwise(x, xp, fp)
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver_tests/ensemble_copula_coupling/test_utilities.py]: fixed, thanks!

File: improver_tests/ensemble_copula_coupling/test_utilities.py [hunk 22]
```diff
@@ -661,6 +661,35 @@ def test_fast_interpolate_pointwise_called(self, interp_imp):
...
+        xp = np.ones((2,))
+        fp = np.ones((2,))
+        msg = r"xp and fp must have at least 2 dimensions"
+        with self.assertRaisesRegex(ValueError, msg):
+            slow_interpolate_pointwise(x, xp, fp)
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply improver_tests/ensemble_copula_coupling/test_utilities.py]: fixed

**Review by gavinevans — APPROVED** [review]
